### PR TITLE
Refactor Gaussian to use information form

### LINKED ISSUE #184: Switch Gaussian to use information filter form, avoiding pseudoinverses
@fehiepsi implemented a `Gaussian` class in Pyro https://github.com/pyro-ppl/pyro/pull/1980 using the information filter form of the Gaussian distribution (i.e. `(info_vec, precision)`). This form obviates the need for pseudoinverses, as e.g. in `funsor.gaussian.sym_solve_mv()`. @fritzo would like to switch to this form, but keep `Gaussian` zero-elevation by omitting the `log_normalizer` term (which in funsor can be stored in the `discrete` component of a `Joint` or similar).

### PR DESCRIPTION
**Pull request by fritzo**
Resolves #184
Addresses #177
Blocking #185 #180 

This refactors the `Gaussian` class to use the "information filter" representation, replacing
```
(loc, precision)  -->  (precision @ loc, precision)
```
This allows `Gaussian` operations on rank-deficient factors, as e.g. arise in `GaussianHMM` #185 and `SwitchingLinearHMM` #180 . This PR also switches from `torch.inverse()` to more numerically stable Cholesky-based inverse and solve operations.

@fehiepsi I have chosen to leave the `0.5 * n * log(2 * pi)` factor out of the log normalizer in this PR, which makes `Gaussian`s closed under addition and leads to convenient identity `∀g:Gaussian, g(0) == 0`. Note that normalizing factors are accounted for in to-funsor conversion, e.g. in `funsor.distributions` and `funsor.pyro.convert.mvn_to_funsor()`.

## Tested
I am fairly confident that this PR is correct, given our good test coverage.
- refactoring is exercised by existing tests
- added a test for the new invariant `g(0) = 0`
- added unit tests for `cholesky_solve` and `cholesky_inverse`

### CODE REVIEW

File: funsor/distributions.py [hunk 1]
```diff
@@ -411,9 +413,10 @@ def eager_normal(loc, scale, value):
             precision[..., i, j] = c1 * c2
     loc = loc.as_tensor()
     precision = precision.as_tensor()
+    info_vec = precision.matmul(loc.unsqueeze(-1)).squeeze(-1)
 
-    log_prob = -0.5 * math.log(2 * math.pi) - scale.log()
-    return log_prob + Gaussian(loc, precision, affine.inputs)
+    log_prob = -0.5 * math.log(2 * math.pi) - scale.data.log() - 0.5 * (loc * info_vec).sum(-1)
```
[the remark this answers is not in the document]
Comment (fritzo) [reply funsor/distributions.py]: Here `scale` is a `funsor.torch.Tensor` and `.data` pulls out its underlying `torch.Tensor`.

[the remark this answers is not in the document]
Comment (fehiepsi) [reply funsor/distributions.py]: That makes sense to me! Previously, log_prob is returned and now, Tensor(log_prob, ...) is returned. I didn't notice it.

**Comment by fritzo** [comment]
@fehiepsi thanks for your careful review!

**Comment by fritzo** [comment]
> I feel that the use of sym_inverse in moment_matching_reduce can be revised

Yes, we will be working on `moment_matching_reduce()` in the future. Right now I've used `sym_inverse()` because I'm not sure whether the original `precision` will be full rank, and this algorithm should work fine with pseudoinverses. I'll see if we can remove the second `sym_inverse()`.

**Pull request merged**